rust/libceed/src/operator.rs

```diff
@@ -20,18 +20,254 @@
+    ///     .basis_tensor_H1_Lagrange(1, 1, 2, q, QuadMode::Gauss)
+    ///     .unwrap();
+    ///
+    /// // Operator fields
+    /// let op = ceed
+    ///     .operator(&qf, QFunctionOpt::None, QFunctionOpt::None)
+    ///     .unwrap()
+    ///     .field("dx", &r, &b, VectorOpt::Active)
+    ///     .unwrap()
+    ///     .field("weights", ElemRestrictionOpt::None, &b, VectorOpt::None)
+    ///     .unwrap()
+    ///     .field("qdata", &rq, BasisOpt::Collocated, VectorOpt::Active)
+    ///     .unwrap();
+    ///
+    /// let inputs = op.inputs().unwrap();
+    ///
+    /// assert_eq!(inputs[0].name(), "dx", "Incorrect input name");
+    /// assert_eq!(inputs[1].name(), "weights", "Incorrect input name");
+    /// ```
+    pub fn name(&self) -> &str {
+        let mut name_ptr: *mut std::os::raw::c_char = std::ptr::null_mut();
+        unsafe {
+            bind_ceed::CeedOperatorFieldGetName(self.ptr, &mut name_ptr);
+        }
+        unsafe { CStr::from_ptr(name_ptr) }.to_str().unwrap()
+    }
+
+    /// Get the ElemRestriction of an OperatorField
+    ///
+    /// ```
+    /// # use libceed::prelude::*;
+    /// # let ceed = libceed::Ceed::default_init();
+    /// let qf = ceed.q_function_interior_by_name("Mass1DBuild").unwrap();
+    ///
+    /// // Operator field arguments
+    /// let ne = 3;
+    /// let q = 4 as usize;
+    /// let mut ind: Vec<i32> = vec![0; 2 * ne];
+    /// for i in 0..ne {
+    ///     ind[2 * i + 0] = i as i32;
+    ///     ind[2 * i + 1] = (i + 1) as i32;
+    /// }
+    /// let r = ceed
+    ///     .elem_restriction(ne, 2, 1, 1, ne + 1, MemType::Host, &ind)
+    ///     .unwrap();
+    /// let strides: [i32; 3] = [1, q as i32, q as i32];
+    /// let rq = ceed
+    ///     .strided_elem_restriction(ne, 2, 1, q * ne, strides)
+    ///     .unwrap();
+    ///
+    /// let b = ceed
+    ///     .basis_tensor_H1_Lagrange(1, 1, 2, q, QuadMode::Gauss)
+    ///     .unwrap();
+    ///
+    /// // Operator fields
+    /// let op = ceed
+    ///     .operator(&qf, QFunctionOpt::None, QFunctionOpt::None)
+    ///     .unwrap()
+    ///     .field("dx", &r, &b, VectorOpt::Active)
+    ///     .unwrap()
+    ///     .field("weights", ElemRestrictionOpt::None, &b, VectorOpt::None)
+    ///     .unwrap()
+    ///     .field("qdata", &rq, BasisOpt::Collocated, VectorOpt::Active)
+    ///     .unwrap();
+    ///
+    /// let inputs = op.inputs().unwrap();
+    ///
+    /// assert!(
+    ///     inputs[0].elem_restriction().is_some(),
+    ///     "Incorrect field ElemRestriction"
+    /// );
+    /// assert!(
+    ///     inputs[1].elem_restriction().is_none(),
+    ///     "Incorrect field ElemRestriction"
+    /// );
+    /// ```
+    pub fn elem_restriction(&self) -> ElemRestrictionOpt {
+        if self.elem_restriction.ptr == unsafe { bind_ceed::CEED_ELEMRESTRICTION_NONE } {
+            ElemRestrictionOpt::None
+        } else {
+            ElemRestrictionOpt::Some(&self.elem_restriction)
+        }
+    }
```

Comment:
Since the purpose of this is to facilitate construction of one operator/qfunction from another, I reworked my approach. Now access is only provided to a *Opt, which contains an immutable ref, so users can only use these fields to construct fields for additional qfunction/operators. They have no direct mutable access to the underlying libCEED object. I think it is now fine to hand the user a vector that they can own because they cannot change the underlying data.

Reply:
We can now, for instance
```rust
// ...
let inputs = op_1.inputs()?;

let op_2 = ceed
     .operator(&qf, QFunctionOpt::None, QFunctionOpt::None)?
     .field(inputs[1].name(), inputs[1].elem_restriction(), inputs[1].basis(), inputs[1].vector())?;
     // ...
```

Reply:
Note: I plan to squash this into the previous commit if we like this approach.